src/FuncSharp/Collections/IEnumerableExtensions_ReadOnlyList.cs

```diff
@@ -1,100 +1,122 @@
+    [Pure]
+    public static T Single<T>(this IReadOnlyList<T> list)
     {
-        if (!source.IsSingle())
-        {
-            throw new ArgumentException("Source is not a single element.");
-        }
-
-        return source[0];
+        return list.IsSingle()
+            ? list[0]
+            : throw new ArgumentException("Source is not a single element.");
     }
 
-    public static IOption<T> SingleOption<T>(this IReadOnlyList<T> source)
+    [Pure]
+    public static IOption<T> SingleOption<T>(this IReadOnlyList<T> list)
     {
-        return source.IsSingle()
-            ? Option.Valued(source[0])
+        return list.IsSingle()
+            ? Option.Valued(list[0])
             : Option.Empty<T>();
     }
 
-    public static T First<T>(this IReadOnlyList<T> source)
+    [Pure]
+    public static T First<T>(this IReadOnlyList<T> list)
     {
-        return source.IsEmpty()
-            ? throw new ArgumentException("Source is empty.")
-            : source[0];
+        return list.ElementAt(0);
     }
 
-    public static IOption<T> FirstOption<T>(this IReadOnlyList<T> source)
+    [Pure]
+    public static IOption<T> FirstOption<T>(this IReadOnlyList<T> list)
     {
-        return source.IsEmpty()
+        return list.IsEmpty()
             ? Option.Empty<T>()
-            : Option.Valued(source[0]);
+            : Option.Valued(list[0]);
+    }
+
+    [Pure]
+    public static T Second<T>(this IReadOnlyList<T> list)
+    {
+        return list[1];
+    }
+
+    public static T Third<T>(this IReadOnlyList<T> list)
+    {
+        return list[2];
+    }
+
+    public static T Fourth<T>(this IReadOnlyList<T> list)
+    {
+        return list[3];
     }
 
-    public static T Second<T>(this IReadOnlyList<T> source)
+    public static T Fifth<T>(this IReadOnlyList<T> list)
     {
-        return source.Count <= 1
-            ? throw new ArgumentException("Source has less than 2 elements.")
-            : source[1];
+        return list[4];
     }
 
-    public static T Last<T>(this IReadOnlyList<T> source)
+    [Pure]
+    public static T Last<T>(this IReadOnlyList<T> list)
     {
-        return source.Count == 0
+        return list.Count == 0
             ? throw new ArgumentException("Source is empty.")
-            : source[source.Count - 1];
+            : list[list.Count - 1];
     }
 
-    public static IOption<T> LastOption<T>(this IReadOnlyList<T> source)
+    [Pure]
+    public static IOption<T> LastOption<T>(this IReadOnlyList<T> list)
     {
-        return source.IsEmpty()
+        return list.IsEmpty()
             ? Option.Empty<T>()
-            : Option.Valued(source[source.Count - 1]);
+            : Option.Valued(list[list.Count - 1]);
     }
 
-    public static T ElementAt<T>(this IReadOnlyList<T> source, int index)
+    [Pure]
+    public static T ElementAt<T>(this IReadOnlyList<T> list, int index)
     {
-        return source[index];
+        return list[index];
     }
```

Comment:
Oh here's the ElementAt, I think this makes the First, Second etc. kind of useless. How about just introducing exception handling here to make sure we throw what we wanna throw?